pkg/filetree/depth_first_path_walker_test.go

```diff
@@ -0,0 +1,287 @@
+		ShouldVisit: func(p file.Path, fn filenode.FileNode) bool {
+			if strings.Contains(string(p), "/home/wagoodman") {
+				return true
+			}
+			return false
+		},
+	}
+
+	walker := NewDepthFirstPathWalker(tr, visitor, &conditions)
+	if err := walker.WalkAll(); err != nil {
+		t.Fatalf("could not walk: %+v", err)
+	}
+
+	assertExpectedTraversal(t, possiblePaths, actualPaths)
+}
+
+func TestDFS_WalkAll_ConditionalBranchPruning(t *testing.T) {
+	tr, possiblePaths := dfsTestTree(t)
+
+	// delete paths we aren't expecting
+	for p := range possiblePaths {
+		if !strings.Contains(p, "/home") && p != "/" && p != "/place" && p != "/sym-linked-dest" && p != "/hard-linked-dest" {
+			delete(possiblePaths, p)
+		}
+	}
+	// </delete>
+
+	actualPaths := make(map[string]*file.Reference, 0)
+	visitor := func(path file.Path, node filenode.FileNode) error {
+		actualPaths[string(path)] = node.Reference
+		return nil
+	}
+
+	conditions := WalkConditions{
+		ShouldContinueBranch: func(p file.Path, fn filenode.FileNode) bool {
+			if p == "/" || strings.Contains(string(p), "/home") {
+				return true
+			}
+			return false
+		},
+	}
+
+	walker := NewDepthFirstPathWalker(tr, visitor, &conditions)
+	if err := walker.WalkAll(); err != nil {
+		t.Fatalf("could not walk: %+v", err)
+	}
+
+	assertExpectedTraversal(t, possiblePaths, actualPaths)
+}
+
+func TestDFS_WalkAll_MaxDirDepthTerminatesTraversal(t *testing.T) {
+	tr := NewFileTree()
+
+	possiblePaths := make(map[string]*file.Reference)
+
+	// absolute symlink
+	_, err := tr.AddSymLink("/home/wagoodman", "/home")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	// since we are following base links on walk, we should NOT expect the symlink ref at the link destination
+	possiblePaths["/home/wagoodman"] = nil
+	possiblePaths["/home"] = nil
+
+	// </setup tree...>
+
+	actualMaxDepth := -1
+	shouldTerminate := func(path file.Path, node filenode.FileNode) bool {
+		if actualMaxDepth > maxDirDepth*2 {
+			// test stop gap
+			t.Fatalf("did not prevent max dir depth traversal")
+			return true
+		}
+		return false
+	}
+
+	visitor := func(path file.Path, node filenode.FileNode) error {
+		actualMaxDepth = strings.Count(string(path.Normalize()), file.DirSeparator)
+		return nil
+	}
+
+	walker := NewDepthFirstPathWalker(tr, visitor, &WalkConditions{
+		ShouldTerminate: shouldTerminate,
+	})
+	if err = walker.WalkAll(); !errors.Is(err, ErrMaxTraversalDepth) {
+		t.Fatalf("expected max traversal error, but got another error instead: %+v", err)
+	} else if err == nil {
+		t.Fatalf("expected max traversal error, but got none")
+	}
+
+	if actualMaxDepth == -1 || actualMaxDepth > maxDirDepth {
+		t.Fatalf("never traversed or went above allowable threshold")
+	}
+
+}
+
+func assertExpectedTraversal(t *testing.T, expected, actual map[string]*file.Reference) {
+	t.Helper()
+	if len(actual) != len(expected) {
+		t.Errorf("Did not traverse all nodes (expected %d, got %d)", len(expected), len(actual))
```

Comment:
tiny nit: the comparison is done with `actual` to `expected` but the messaging is `expected` to `actual`. I don't have a strong preference but consistency would be nice, especially when there are failures